plugins/filter_watermark/watermark.c

```diff
@@ -0,0 +1,385 @@
+         /*Dump all data into array to flush out*/
+         tmp_record = c_heap_read_root(h);
+         if (tmp_record == NULL) {
+             /* This place could not be reached, since at least one piece of record is in the heap*/
+             flb_plg_error(context->ins,  "Fatal Error!");
+             ret = -1;
+             break;
+         }
+         tmp_record_timestamp = &(tmp_record->time_stamp);
+         strftime(outstr, sizeof(outstr), fmt, tmp_record_timestamp);
+         flb_plg_debug(context->ins, "%s        %ld\n", outstr, mktime(tmp_record_timestamp));
+
+         /* Make a copy of ptr of record and pop the heap top*/
+         if (mktime(tmp_record_timestamp) <= win_right_edge) { 
+             if (context->record_count >= CACHE_SIZE_DEFAULT_VALUE) {
+                 flb_plg_info(context->ins, "Cache is full and drop the record to flush out\n");
+             } else {
+                 context->record_pointer_cache_out_array[context->record_count] = tmp_record;
+                 ++(context->record_count);
+             }
+             c_heap_get_root(h);
+             ret = 1;
+             continue;
+         } else {
+             /* It means that current top of head is in another windows. */
+             /* Update the left and right edge with new timestamp. */
+             win_left_edge = mktime(tmp_record_timestamp);
+             win_right_edge = win_left_edge + window_size; 
+             break;
+         }
+    }
+  
+    /* Update context. */
+    context->win_right_edge = win_right_edge;
+    context->win_left_edge = win_left_edge;
+    return ret;
+}
+
+
+static int cb_watermark_init(struct flb_filter_instance *f_ins,
+                             struct flb_config *config,
+                             void *data)
+{
+    int ret;
+    struct flb_filter_watermark_ctx *ctx;
+
+    /* Create context */
+    ctx = flb_malloc(sizeof(struct flb_filter_watermark_ctx));
+    if (!ctx) {
+        flb_errno();
+        return -1;
+    }
+    ctx->ins = f_ins;
+
+    /* parse plugin configuration  */
+    ret = configure(ctx, f_ins);
+    if (ret == -1) {
+        flb_free(ctx);
+        return -1;
+    }
+
+    /* Set our context */
+    flb_filter_set_context(f_ins, ctx);
+    ctx->h = c_heap_create(compare_imp, deconstructor_imp);
+    ctx->record_count = 0; 
+    
+    return 0;
+}
+
+static int cb_watermark_filter(const void *data, size_t bytes,
+                               const char *tag, int tag_len,
+                               void **out_buf, size_t *out_size,
+                               struct flb_filter_instance *f_ins,
+                               void *context,
+                               struct flb_config *config)
+{
+    msgpack_object root;
+    msgpack_object map;
+    msgpack_object key;
+    msgpack_object value;
+    int i; 
+    int map_size; 
+    int ret;
+    struct tm tm0;
+
+    msgpack_unpacked result;
+    msgpack_unpacked * destroy_result;
```

Comment:
`*destroy_result`

Reply:
fixed in commit https://github.com/fluent/fluent-bit/commit/4b46f5c1048639ed37cafc345e575f6794dccba5